arangod/Aql/Optimizer/Rule/OptimizerRulePushLimitIntoIndex.cpp

```diff
@@ -0,0 +1,242 @@
+    }
+
+    // condition of IndexNode can only be a single `compare in` for the
+    // rule to be applicable
+    // the tree has a specific format => DNF therefore the assumption
+    // of having only one member and expecting COMPARE IN node
+    // is valid
+    auto const* compareInNode = indexNode->condition() == nullptr
+                                    ? nullptr
+                                    : indexNode->condition()->root();
+    while (compareInNode != nullptr) {
+      if (compareInNode->type == NODE_TYPE_OPERATOR_BINARY_IN) {
+        break;
+      }
+      if (compareInNode->numMembers() != 1) {
+        compareInNode = nullptr;
+        break;
+      }
+
+      compareInNode = compareInNode->getMember(0);
+    }
+    if (compareInNode == nullptr) {
+      continue;
+    }
+
+    auto const& indexes = indexNode->getIndexes();
+    if (indexes.size() != 1) {
+      // IndexNode uses more than a single index. not safe to apply the
+      // optimization here.
+      continue;
+    }
+    auto const& usedIndex = indexes.front();
+    if (!isEligibleIndex(usedIndex)) {
+      continue;
+    }
+
+    // remember the output variable that is produced by the IndexNode
+    Variable const* outVariable = indexNode->outVariable();
+
+    // check if `compare in` variable is same as IndexNode output,
+    // and that it compares against the attribute defined in index
+    if (auto const* lhs = compareInNode->getMember(0);
+        lhs->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
+      std::pair<Variable const*, std::vector<arangodb::basics::AttributeName>>
+          attributeAccessResult;
+      if (!lhs->isAttributeAccessForVariable(attributeAccessResult, false)) {
+        continue;
+      }
+      if (attributeAccessResult.first != outVariable) {
+        continue;
+      }
+      if (attributeAccessResult.second != *usedIndex->fields().begin()) {
+        continue;
+      }
+    } else {
+      continue;
+    }
+
+    // check if index node parents contains a pair of sort and limit
+    auto const* sortNode = indexNode->getFirstParent();
+    while (sortNode != nullptr && sortNode->getType() == EN::CALCULATION) {
+      sortNode = sortNode->getFirstParent();
+    }
+
+    if (sortNode == nullptr || sortNode->getType() != EN::SORT) {
+      continue;
+    }
+
+    auto const& sortFields =
+        ExecutionNode::castTo<SortNode const*>(sortNode)->elements();
+
+    if (!isEligibleSort(usedIndex->fields().begin(), usedIndex->fields().end(),
+                        sortFields, plan.get(), indexNode) &&
+        !isEligibleSort(usedIndex->fields().begin() + 1,
+                        usedIndex->fields().end(), sortFields, plan.get(),
+                        indexNode)) {
+      continue;
+    }
+
+    ExecutionNode* limitNode = sortNode->getFirstParent();
+    if (limitNode == nullptr || limitNode->getType() != EN::LIMIT) {
+      continue;
+    }
+
+    indexNode->setLimit(ExecutionNode::castTo<LimitNode*>(limitNode)->limit());
```

Comment:
I think this only works if we have `LIMIT x` or `LIMIT 0, x`.
If we have `LIMIT y, x` (with y > 0), then the limit we need to push forward is `y + x`.
We should also add tests for this.
```suggestion
    indexNode->setLimit(ExecutionNode::castTo<LimitNode*>(limitNode)->offset() + ExecutionNode::castTo<LimitNode*>(limitNode)->limit());
```

Reply:
Ah yes, I have forgotten about this! Thanks